In [4]:
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt

In [5]:
url = "https://raw.githubusercontent.com/lopez-isaac/project-data-sets/master/2019%20Winter%20Data%20Science%20Intern%20Challenge%20Data%20Set%20-%20Sheet1.csv"
shoe_sales = pd.read_csv(url)

In [6]:
print(shoe_sales.shape)
shoe_sales.head()

(5000, 7)


order_id  shop_id  user_id  ...  total_items  payment_method           created_at
0         1       53      746  ...            2            cash  2017-03-13 12:36:56
1         2       92      925  ...            1            cash  2017-03-03 17:38:52
2         3       44      861  ...            1            cash   2017-03-14 4:23:56
3         4       18      935  ...            1     credit_card  2017-03-26 12:43:37
4         5       18      883  ...            1     credit_card   2017-03-01 4:35:11

[5 rows x 7 columns]

In [7]:
#average order amount (AOV)
shoe_sales["order_amount"].mean()

3145.128

In [8]:
#quick describtive statistics of the column in question
#The the max value in the column is $704000
shoe_sales["order_amount"].describe()

count      5000.000000
mean       3145.128000
std       41282.539349
min          90.000000
25%         163.000000
50%         284.000000
75%         390.000000
max      704000.000000
Name: order_amount, dtype: float64

In [9]:
shoe_sales.sort_values(by=["order_amount"])
#Looking for the outlier. We see that its not some expensive shoe but large large transaction orders.

order_id  shop_id  ...  payment_method          created_at
158        159       92  ...     credit_card  2017-03-29 3:07:12
3871      3872       92  ...           debit  2017-03-18 9:10:08
4760      4761       92  ...           debit  2017-03-20 7:37:28
4923      4924       92  ...     credit_card  2017-03-09 5:05:11
4932      4933       92  ...     credit_card  2017-03-24 2:17:13
...        ...      ...  ...             ...                 ...
2153      2154       42  ...     credit_card  2017-03-12 4:00:00
1562      1563       42  ...     credit_card  2017-03-19 4:00:00
1362      1363       42  ...     credit_card  2017-03-15 4:00:00
520        521       42  ...     credit_card  2017-03-02 4:00:00
3332      3333       42  ...     credit_card  2017-03-24 4:00:00

[5000 rows x 7 columns]

In [10]:
# How many pairs are being bought at once and how oftan
shoe_sales['total_items'].value_counts(sort=True, ascending=False)

2       1832
1       1830
3        941
4        293
5         77
2000      17
6          9
8          1
Name: total_items, dtype: int64

#First glance analysis


*   The reason for the AOV being $3145 is due to large transaction orders.

* However those large transaction orders are not being made oftan customers are being 2 or 1 shoes at a time.

* Who is making these large transactions? How oftan? 




In [11]:
big_spender_df = shoe_sales.loc[shoe_sales['order_amount'] == 704000]

In [12]:
#Same user and sane store is making the repeated large transaction
print(big_spender_df.shape)
big_spender_df.head(18)

(17, 7)


order_id  shop_id  ...  payment_method          created_at
15          16       42  ...     credit_card  2017-03-07 4:00:00
60          61       42  ...     credit_card  2017-03-04 4:00:00
520        521       42  ...     credit_card  2017-03-02 4:00:00
1104      1105       42  ...     credit_card  2017-03-24 4:00:00
1362      1363       42  ...     credit_card  2017-03-15 4:00:00
1436      1437       42  ...     credit_card  2017-03-11 4:00:00
1562      1563       42  ...     credit_card  2017-03-19 4:00:00
1602      1603       42  ...     credit_card  2017-03-17 4:00:00
2153      2154       42  ...     credit_card  2017-03-12 4:00:00
2297      2298       42  ...     credit_card  2017-03-07 4:00:00
2835      2836       42  ...     credit_card  2017-03-28 4:00:00
2969      2970       42  ...     credit_card  2017-03-28 4:00:00
3332      3333       42  ...     credit_card  2017-03-24 4:00:00
4056      4057       42  ...     credit_card  2017-03-28 4:00:00
4646      4647       42  ...     credit_card  2017-03-02 4:00:00
4868      4869       42  ...     credit_card  2017-03-22 4:00:00
4882      4883       42  ...     credit_card  2017-03-25 4:00:00

[17 rows x 7 columns]

In [13]:
#How oftan?
#There isn't specific pattern, but a purchase is made within at least 4 days
big_spender_df.sort_values(by=["created_at"])

order_id  shop_id  ...  payment_method          created_at
520        521       42  ...     credit_card  2017-03-02 4:00:00
4646      4647       42  ...     credit_card  2017-03-02 4:00:00
60          61       42  ...     credit_card  2017-03-04 4:00:00
15          16       42  ...     credit_card  2017-03-07 4:00:00
2297      2298       42  ...     credit_card  2017-03-07 4:00:00
1436      1437       42  ...     credit_card  2017-03-11 4:00:00
2153      2154       42  ...     credit_card  2017-03-12 4:00:00
1362      1363       42  ...     credit_card  2017-03-15 4:00:00
1602      1603       42  ...     credit_card  2017-03-17 4:00:00
1562      1563       42  ...     credit_card  2017-03-19 4:00:00
4868      4869       42  ...     credit_card  2017-03-22 4:00:00
3332      3333       42  ...     credit_card  2017-03-24 4:00:00
1104      1105       42  ...     credit_card  2017-03-24 4:00:00
4882      4883       42  ...     credit_card  2017-03-25 4:00:00
2835      2836       42  ...     credit_card  2017-03-28 4:00:00
2969      2970       42  ...     credit_card  2017-03-28 4:00:00
4056      4057       42  ...     credit_card  2017-03-28 4:00:00

[17 rows x 7 columns]


* Do All the shoes cost them same
*   What store Selling the most items?
*   What store is making the most revanue?



In [14]:
# Do all the shoes cost the same
shoe_price = shoe_sales.loc[shoe_sales['total_items'] == 1].drop_duplicates(subset=["shop_id"])

In [15]:
shoe_price.head()

order_id  shop_id  user_id  ...  total_items  payment_method           created_at
1         2       92      925  ...            1            cash  2017-03-03 17:38:52
2         3       44      861  ...            1            cash   2017-03-14 4:23:56
3         4       18      935  ...            1     credit_card  2017-03-26 12:43:37
5         6       58      882  ...            1     credit_card  2017-03-14 15:25:01
6         7       87      915  ...            1            cash  2017-03-01 21:37:57

[5 rows x 7 columns]

In [16]:
#range of prices
#cheapest shoe is $90, highest is $25725
shoe_price['order_amount'].describe()

count      100.000000
mean       407.990000
std       2557.462906
min         90.000000
25%        132.750000
50%        153.000000
75%        168.250000
max      25725.000000
Name: order_amount, dtype: float64

In [17]:
#what shops are selling this expensive shoe
shoe_price.loc[shoe_price['order_amount'] == 25725]

order_id  shop_id  user_id  ...  total_items  payment_method          created_at
160       161       78      990  ...            1     credit_card  2017-03-12 5:56:57

[1 rows x 7 columns]

In [18]:
#dataframe used to do inner join on main dataset
shoe_price = shoe_price.rename(columns={"order_amount":"shoe_value"})
shoe_price_merge_df = shoe_price[["shoe_value", "shop_id"]]
shoe_price_merge_df.head()

shoe_value  shop_id
1          90       92
2         144       44
3         156       18
5         138       58
6         149       87

In [19]:
#add this featured column to dataset
shoe_sales = pd.merge(shoe_sales,shoe_price_merge_df,on='shop_id')
shoe_sales.head()

order_id  shop_id  user_id  ...  payment_method           created_at shoe_value
0         1       53      746  ...            cash  2017-03-13 12:36:56        112
1        28       53      942  ...     credit_card   2017-03-17 9:41:53        112
2        57       53      739  ...     credit_card   2017-03-18 8:45:21        112
3        86       53      908  ...     credit_card  2017-03-27 20:46:35        112
4       104       53      740  ...            cash   2017-03-17 7:29:30        112

[5 rows x 8 columns]

In [23]:
pd.options.display.max_rows = 150

In [46]:
#are customers buying cheaper shoes more oftan than expensive shoes?
shoe_sales.groupby(["shoe_value", "shop_id"])["total_items"].sum().sort_values(ascending=False)

shoe_value  shop_id
352         42         34063
160         13           136
153         84           132
164         71           130
112         53           130
177         81           128
118         41           127
163         19           126
140         24           126
187         6            121
178         59           121
116         14           121
131         69           121
148         10           119
163         29           118
196         89           118
165         73           118
173         70           117
118         9            117
154         65           116
142         37           115
129         49           115
136         63           113
156         18           112
            23           112
130         86           111
114         93           111
178         90           111
160         91           110
142         45           110
153         96           110
138         58           109
112         7            109
158         61         

#Final analysis

##b.What metric would you report for this dataset?

I would use the total items column. More specifically the frequency of how many pairs are being purchased at once. From that you could create promotions involving multiple shoe pair purchases for both consumer and business. Single purchase transactions do not outnumber the sum of all multy pair purchases. 
##c.What is its value?




In [35]:
# Frequency of how many pairs are being bought at once
shoe_sales['total_items'].value_counts(sort=True, ascending=False)

2       1832
1       1830
3        941
4        293
5         77
2000      17
6          9
8          1
Name: total_items, dtype: int64

#How many orders were shipped by Speedy Express in total?

###SQL statement
SELECT COUNT(shipperid) FROM [Orders]

WHERE shipperid = 1;

#ANSWER: 
54


#What is the last name of the employee with the most orders?

###SQL statement

SELECT employeeid, lastname, COUNT(EmployeeID) 

FROM (SELECT * FROM orders

INNER join employees
	


> on orders.employeeid = employees.employeeid)



GROUP BY employeeid

ORDER BY count(employeeid) DESC

LIMIT 1


#ANSWER:
EmployeeID: 4	
LastName: Peacock	
COUNT(EmployeeID):40

#What product was ordered the most by customers in Germany?

###SQL statement

SELECT Country, ProductID, SUM(Quantity), Productname

FROM ( SELECT * FROM Orders

INNER JOIN Customers
> ON Orders.customerID = Customers.CustomerID



INNER JOIN Orderdetails

> ON Orders.OrderID = Orderdetails.OrderID


INNER JOIN Products

> ON Orderdetails.ProductID = Products.ProductID)

WHERE Country LIKE "germany"

GROUP BY ProductID

ORDER BY SUM(Quantity) DESC


#ANSWER:
	
Country: Germany

ProductID: 40

SUM(Quantity):	160

ProductName: Boston Crab Meat

